### PR TITLE
Cloud provider refactor

### PR DESCRIPTION
**Pull request by ash2k**
The idea is to split `CloudHandler` into two pieces by introducing an interface in the middle.
Currently `CloudHandler` does these things:
1. maintains a (negative) cache of ip->instance info
1. maintains a map of ip->[]metrics that need to be enriched per ip while the lookup it taking place

I'm introducing `CachedInstances` interface to abstract the caching bit (1 in the list above) and leave only 2 in the `CloudHandler`.

The next step/PR is to refactor the k8s cloud provider into an implementation of `CachedInstances`.

This PR is not ready to review in full (tests don't even compile as I haven't touched them yet) but to give you a preview and collect early feedback on the approach and major impl details. WDYT?

### CODE REVIEW

**Comment by mwhittington21** [comment]
Seems sane to me! Splitting out the caching behaviour from a default cloud provider makes a lot of sense.

**Comment by ash2k** [comment]
I'll get back to this PR, hopefully soon.

File: pkg/cachedinstances/factory.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/atlassian/gostatsd"
+	"github.com/atlassian/gostatsd/pkg/cachedinstances/cloudprovider"
+	"github.com/atlassian/gostatsd/pkg/cloudproviders/aws"
+	"github.com/atlassian/gostatsd/pkg/cloudproviders/k8s"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"golang.org/x/time/rate"
+)
+
+var (
```
Comment (ash2k) [inline pkg/cachedinstances/factory.go:16]: Two types of defaults will be gone once k8s provider is no longer a provider but a direct implementation of `CachedInstances` interface (next PR).

File: pkg/statsd/statsd.go [hunk 2]
```diff
@@ -88,12 +86,6 @@ func socketFactory(metricsAddr string, connPerReader bool) SocketFactory {
 func (s *Server) createStandaloneSink() (gostatsd.PipelineHandler, []gostatsd.Runnable, error) {
 	var runnables []gostatsd.Runnable
 
-	for _, backend := range s.Backends {
```
Comment (ash2k) [inline pkg/statsd/statsd.go:91]: Server receives backends as a parameter - it is not the owner, it didn't create them. So it should not be managing their lifecycle. That is why I moved this up one level into whatever constructs the `Server`. Same applies to `CachedInstances` and cloud providers.

File: types.go [hunk 3]
```diff
@@ -48,11 +48,25 @@ type TimerSubtypes struct {
...
 type Runner interface {
-	Run(ctx context.Context)
+	Run(context.Context)
+}
+
+type MetricsRunner interface {
+	RunMetricsContext(context.Context)
+}
+
+func MaybeAppendRunnable(runnables []Runnable, maybeRunner interface{}) []Runnable {
```
Comment (ash2k) [inline types.go:62]: Single helper func to ensure we always start both the main work loop and the metrics loop too.

**Comment by ash2k** [comment]
This is ready for review.

**Comment by tiedotguy** [comment]
Awesome! I'll try and give it a thorough review tomorrow.

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+			toReturnInfo = gostatsd.InstanceInfo{} // enable GC
+			toReturnInfoC = nil                    // info has been sent; if there is nothing to send, the case is disabled
+		case info := <-ownInfoSource:
+			ch.handleInstanceInfo(info)
+		case t := <-refreshTicker.C:
+			ch.doRefresh(t)
+		case statser := <-ch.emitChan:
+			ch.emit(statser)
+		}
+		if toLookupC == nil && len(ch.toLookupIPs) > 0 {
```
Comment (tiedotguy) [inline pkg/cachedinstances/cloudprovider/cached_cloud_provider.go:100]: `ch.toLookupIPs` is only ever populated by `ch.doRefresh()`.  `ch.doRefresh()` could just return a slice of IPs to refresh, then this could pass that entire thing as a single slice to `ch.ipSinkSource`, rather than feeding it through one at a time.

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+		case toLookupC <- toLookupIP:
+			toLookupIP = gostatsd.UnknownIP // enable GC
+			toLookupC = nil                 // ip has been sent; if there is nothing to send, the case is disabled
+		case toReturnInfoC <- toReturnInfo:
+			toReturnInfo = gostatsd.InstanceInfo{} // enable GC
+			toReturnInfoC = nil                    // info has been sent; if there is nothing to send, the case is disabled
+		case info := <-ownInfoSource:
+			ch.handleInstanceInfo(info)
+		case t := <-refreshTicker.C:
+			ch.doRefresh(t)
```
Comment (tiedotguy) [inline pkg/cachedinstances/cloudprovider/cached_cloud_provider.go:96]: I would suggest skipping refreshing if we haven't managed to clear the previous refresh yet. I could see a situation where we get in to a bad state (due to cpu or network, not necessarily a bug), where we're refreshing faster than we can process, causing either unbound growth in `ch.toLookupIPs` or having something to write before we're ready if the below suggestion is done.

I'm not so sure if the same pattern is useful in `toReturnInfo`, it complicates the interface a bunch.

tl;dr backpressure on `ch.toLookupIPs`

Comment (ash2k) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go:96]: I agree and I thought about that. But I think it should be done separately. This PR does not make it better or worse, right? Also, don't want to make it even bigger.

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go (hunk 4 above)
Comment (ash2k) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go:100]: What are the benefits? This will complicate logic quite a bit. `cloudProviderLookupDispatcher.run()` would need to be handle all the extra complexity of collecting just enough IPs in a buffer and then processing them. I tried such approach somewhere during this refactoring and found that it complicates things, hurts readability. And when things are not simple, bugs creep in. :)

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	case <-ctx.Done():
+		// success-ish
+	default:
+		// at least we tried
+	}
+}
+
+func (ch *CachedCloudProvider) emit(statser stats.Statser) {
+	// regular
+	statser.Gauge("cloudprovider.cache_positive", float64(ch.statsCachePositive), nil)
```
[the remark this answers is not in the document]
Comment (ash2k) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go:171]: Oh, yeah, I'd like to discuss the metrics we emit as I want' paying too much attention to it while refactoring. Was too much to handle already so I might have disrupted things a bit.
I think not all active components emit stats properly. Some didn't have their `RunMetricsWithContext()`/etc called, some probably never get a non-null statser. An audit should take place. I tried to patch up things a bit, but haven't verified every single component emits stats properly.

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go (hunk 4 above)
Comment (tiedotguy) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go]: Splitting this in to two aspects, which should be considered individually - returning a slice, and passing a slice to the lookup.

Returning a slice removes the extra state from the structure, simplifying the query of "where is this referenced?". Reducing the scope to a local variable means I don't need to have to think about locking, or what else might touch it, because I can see it's a regular local variable, returned from a function.  The only thing that it simplifies currently is no back pressure. If back pressure were added so we didn't call `doRefresh`, then it wouldn't even add that, we'd just be returning a slice through the struct.

Passing a slice to lookup means the lookup dispatcher has to do a little extra thinking, but I wouldn't say it's much. Without writing code, I'd probably wrap it up in a type which has 3 methods... "push more", "pop maxLookupIps or nil", and "get maxLookupIPs or what you have". No channels, no timers, just a simple data structure (probably just a slice) which is trivial to throw in to a test. The actual consumption of the data structure is also easier to follow at that point.

**Comment by ash2k** [comment]
@tiedotguy thanks for the review! I know I've done a few too many changes here that could all be separate PRs/commits. I know it's a poor excuse, but I don't have much time to rework this just for the sake of nicer history (which I do like to have too!). If nobody feels strongly about it, can we merge this (assuming everything else is fine)? If there are objections, I'll do more work here.

Regarding putting things into the root package - I don't mind having them there but also don't mind moving them somewhere else if that's what you and others prefer. I like to have api types there because a hypothetical `gostatsd.Metric`/etc reads nice. In the particular case of this PR I moved things because there was some dependency cycle. I do not remember now what it was, sorry, this PR is "spread in time" too much and I'm forgetting "whats" and "whys".

If we go ahead with this, I can squash it before merging.

Next step, which I'm planning to work on, is to refactor the k8s cloud provider into the new interface and do some cleanup around it (e.g. no two sets of defaults will be needed).

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go (hunk 5 above)
Comment (tiedotguy) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go]: Works for me 👍

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go (hunk 4 above)
Comment (ash2k) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go]: Thanks for elaborating on the idea. I think actually trying it will help to see if it's worth it or not. I think if more logic needs to be packed into how IPs are handled then it may make sense. Another thing I'm thinking of is turning it into a map to dedupe IPs that have been in the queue for too long. And then we can just have a capacity limit on that map. Or maybe we don't even need that capacity limit with deduplication...

File: pkg/cachedinstances/cloudprovider/cached_cloud_provider.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tiedotguy) [reply pkg/cachedinstances/cloudprovider/cached_cloud_provider.go]: I'm sure @adampetrovic or @MovieStoreGuy will let us know if it broke anything :)

In terms of metric changes in this PR, I don't believe there's any loss of fidelity - there's the CP state (cache), and CP actions (actual CP work), and they're just split out, so it's all good.  Maybe for k8s, write the PR without metrics, and then raise an issue to refactor metrics later?

Once we have k8s, we can assess what metrics it can sensible generate, how they overlap with what we already have, and if we need new names / tags / whatever.

**Comment by tiedotguy** [comment]
I did track down the dependency cycle, and I agree it was needed. I don't recall specifically what it was now though :)

My main objection to the root is really the physical file placement. The mixing of source files and project files is a mixing of concerns.  Either way, it doesn't need to be handled in this PR

I think this is clear to merge now.

**Comment by ash2k** [comment]
Cool! I have rebased and squashed, tick it one more time, please.
